[torch][ao] Repro assigning conv2d debug handle to weights and bias

dulinriley opened the pull request:
Summary:
Test related to: https://github.com/pytorch/pytorch/issues/134778

I found a test case that shows that re-exporting the same model again
moves debug handles in surprising ways, for example:
the conv2d debug handle moved from just the conv2d, to the weights
and biases.

If you enable a quantizer like XNNPackQuantizer, it still happens for conv + bias,
but not for the weights.

Test Plan: This isn't a real test, but is the basis for an investigation as to why this occurs and how to fix it

Differential Revision: D62007054

dulinriley commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -188,3 +189,53 @@ def test_extract_results_from_loggers(self):
+            ):
+                quant_handle_to_nodes[n.meta[CUSTOM_KEY][NUMERIC_DEBUG_HANDLE_KEY]].append(n)
+        # This is the bug, bias and conv have the same debug handle.
+        self.assertEqual(quant_handle_to_nodes[3][0].name, "p_getattr_l__self___conv___0___weight")
+        self.assertEqual(quant_handle_to_nodes[3][1].name, "p_getattr_l__self___conv___0___bias")
+        self.assertEqual(quant_handle_to_nodes[3][2].name, "conv2d")
```
The bug is here, it should only have a single node for handle 3, and that should be conv2d. The weight and bias shouldn't have a handle at all (they're placeholders)

yiming0416 commented on test/quantization/pt2e/test_numeric_debugger.py:
```diff
@@ -188,3 +189,53 @@ def test_extract_results_from_loggers(self):
+                and NUMERIC_DEBUG_HANDLE_KEY in n.meta[CUSTOM_KEY]
+            ):
+                ref_handle_to_nodes[n.meta[CUSTOM_KEY][NUMERIC_DEBUG_HANDLE_KEY]].append(n)
+
```
Upon initial investigation, I feel like the issue stems from calling the `.module()` of an `ExportedProgram`

If you don't do the export again. Just print the `ref_handle_to_nodes` here. You will still see the weight and bias of conv written to the handle as shown below 
```
defaultdict(<class 'list'>, {0: [conv_0_weight], 1: [conv_0_bias], 2: [conv2d], 3: [relu]})
```

However, if just let `m = ref_exported`, then the resulting `ref_handle_to_nodes` will be 

```
defaultdict(<class 'list'>, {0: [conv2d], 1: [relu]})
```
Somehow, calling .module() make a `placeholder` becomes a `get_attr`

dulinriley commented:
Sorry @yiming0416 I didn't quite follow your comments, is this fixed in https://github.com/pytorch/pytorch/pull/135720?
So if I wait for that PR to merge, then the test I introduce here will no longer associate the same handle with weight and bias?

yiming0416 commented:
@dulinriley https://github.com/pytorch/pytorch/pull/135720 may need more time before merging as it could break other tests. 

For your test, the debug handle will always be attached to weight and bias if you call `generate_numeric_debug_handle()` on `ep.module()`, since `.module()` unlifts `placeholder` to `get_attr`.

With the fix, we can only guarantee the debug handle of weight and bias will not be changed after re-export.

cc @jerryzh168

dulinriley commented:
So I should *never* call generate_numeric_debug_handle on the `.module()`, but only on the ExportedProgram?
The problem is that we can't run forward on ExportedProgram, and we can't add `call_module` nodes to that, which is what our loggers use.

Should we do something else, like use fx.Interpreter instead?

yiming0416 commented:
@dulinriley I think this question is currently beyond my knowledge scope. I feel like that's more a question on how you should define the `generate_numeric_debug_handle()` function.

I saw most of the quantization tests call `generate_numeric_debug_handle` on `.module()`. Also what I learned from Jerry is that we care about the debug handle should be consistent after re-exporting and it is okay for weight and bias to have debug handles. cc @jerryzh168 can you confirm this? Thanks

dulinriley committed:
[torch][ao] Repro assigning conv2d debug handle to weights and bias (#134864)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/134864

Test related to: https://github.com/pytorch/pytorch/issues/134778

I found a test case that shows that re-exporting the same model again
moves debug handles in surprising ways, for example:
the conv2d debug handle moved from just the conv2d, to the weights
and biases.

If you enable a quantizer like XNNPackQuantizer, it still happens for conv + bias,
but not for the weights.

Test Plan: This isn't a real test, but is the basis for an investigation as to why this occurs and how to fix it

Differential Revision: D62007054
